torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c

```diff
@@ -0,0 +1,302 @@
+  {
+    THTensor_(resize4d)(output, nslices, osizeT, osizeH, osizeW);
+
+    input_data = THTensor_(data)(input);
+    output_data = THTensor_(data)(output);
+
+    THNN_(VolumetricAdaptiveAveragePooling_updateOutput_frame)(input_data, output_data,
+                                                      nslices,
+                                                      isizeT, isizeW, isizeH,
+                                                      osizeT, osizeW, osizeH,
+                                                      istrideT, istrideW, istrideH,
+                                                      istrideD);
+  }
+  else
+  {
+    long p;
+
+    THTensor_(resize5d)(output, nbatch, nslices, osizeT, osizeH, osizeW);
+
+    input_data = THTensor_(data)(input);
+    output_data = THTensor_(data)(output);
+
+#pragma omp parallel for private(p)
+    for (p = 0; p < nbatch; p++)
+    {
+      THNN_(VolumetricAdaptiveAveragePooling_updateOutput_frame)(input_data+p*istrideB, output_data+p*nslices*osizeT*osizeW*osizeH,
+                                                        nslices,
+                                                        isizeT, isizeW, isizeH,
+                                                        osizeT, osizeW, osizeH,
+                                                        istrideT, istrideW, istrideH,
+                                                        istrideD);
+    }
+  }
+}
+
+static void THNN_(VolumetricAdaptiveAveragePooling_updateGradInput_frame)(
+          real *gradInput_p,
+          real *gradOutput_p,
+          long nslices,
+          long isizeT,
+          long isizeW,
+          long isizeH,
+          long osizeT,
+          long osizeW,
+          long osizeH)
+{
+  long k;
+#pragma omp parallel for private(k)
+  for (k = 0; k < nslices; k++)
+  {
+    real *gradInput_p_k = gradInput_p + k*isizeT*isizeW*isizeH;
+    real *gradOutput_p_k = gradOutput_p + k*osizeT*osizeW*osizeH;
+
+    /* calculate average */
+    long ti, i, j;
+    for(ti = 0; ti < osizeT; ti++)
+    {
+      int z_start = START_IND(ti, osizeT, isizeT);
+      int z_end   = END_IND(ti, osizeT, isizeT);
+      int kT = z_end - z_start;
+
+      for(i = 0; i < osizeH; i++)
+      {
+        int y_start = START_IND(i, osizeH, isizeH);
+        int y_end   = END_IND(i, osizeH, isizeH);
+        int kH = y_end - y_start;
+
+        for(j = 0; j < osizeW; j++)
+        {
+
+          int x_start = START_IND(j, osizeW, isizeW);
+          int x_end   = END_IND(j, osizeW, isizeW);
+          int kW = x_end - x_start;
+
+          real grad_delta = gradOutput_p_k[ti*osizeH*osizeW + i*osizeW + j] / kT / kW / kH;
+
+          int x, y, z;
+          for(z = z_start; z < z_end; z++)
+          {
+            for(y = y_start; y < y_end; y++)
+            {
+              for(x = x_start; x < x_end; x++)
+              {
+                /* update gradient */
+                gradInput_p_k[z*isizeH*isizeW + y*isizeW + x] += grad_delta;
```

Comment:
I'm suddenly confused by this line (and the corresponding line in spatial version). Why is this not atomic?

Reply:
dont rule out bugs :)
It does indeed look like a race

Reply:
Because it's parallelized batch and channel dimension.

Reply:
Thanks for enlightening me, @ruotianluo !